docs: haw-sequencer (#3367)
OffchainLabs/arbitrum-docs: docs/how-arbitrum-works/deep-dives/finality.mdx

Arbitrum offers two levels of finality:

1. **Soft finality**: Provided by the Sequencer’s real-time feed, offering immediate but provisional transaction confirmations.
2. **Hard finality**: Occurs when transactions are included in batches posted to and finalized on the parent chain, providing strong security assurances. Final settlement on the parent chain happens via [assertions](/how-arbitrum-works/deep-dives/assertions.mdx) confirmed by the Rollup protocol.

## Soft finality

Soft finality refers to the preliminary confirmation of transactions based on the Sequencer’s real-time feed. Key characteristics include:

- **Immediate confirmation**: Transactions are confirmed almost instantly as they are accepted and ordered by the Sequencer.
- **Provisional assurance**: The confirmations are provisional and rely on the Sequencer’s integrity and availability.
- **High performance**: Enables applications to offer rapid responses and real-time interactions, enhancing user experience.

> #### Advantages of soft finality

Reviewer: ```suggestion
### Advantages of soft finality
```

Author: For these headings—I know I'm technically breaking nesting rules. However, with the length of doc and headings, I'm trying to minimize clutter in sidebar (right).

- **Low latency**: Users receive immediate feedback on transaction status.
- **Optimized for speed**: Ideal for applications where responsiveness is critical.
- **Improved user experience**: Reduces waiting times and uncertainty.

> #### Limitations of soft finality

Reviewer: ```suggestion
### Limitations of soft finality
```

- **Trust dependency**: Relies on the Sequencer’s honesty and ability to maintain uptime.
- **Potential for reordering**: In rare cases, if the Sequencer acts maliciously or encounters issues, the provisional ordering could change.
- **Not suitable for high-value transactions**: For transactions requiring strong security guarantees, soft finality may not suffice.

## Hard finality

> Hard finality occurs when batched transactions get posted to the parent chain. Key characteristics include:

Reviewer: ```suggestion
Hard finality occurs when batched transactions are posted to the parent chain. Key characteristics include:
```

- **Strong security guarantees**: When included in blocks on the parent chain, transactions inherit the parent chain’s security assurances.
- **Irreversibility**: Once finalized, transactions are immutable and cannot be altered or reversed.
- **Data availability**: All transaction data is recorded onchain, ensuring transparency and verifiability.

> #### Advantages of hard finality

Reviewer: ```suggestion
### Advantages of hard finality
```

- **Maximum security**: Protected by the robustness of the parent chain’s consensus mechanism.

> - **Trust minimization**: This does not require trust in the Sequencer; the underlying blockchain provides security.

Reviewer: ```suggestion
- **Trust minimization**: Does not require trust in the Sequencer; the underlying blockchain provides security.
```

- **Suitable for high-value transactions**: Ideal for scenarios where security and immutability are paramount.

> #### Limitations of hard finality

Reviewer: ```suggestion
### Limitations of hard finality
```

- **High latency**: Achieving hard finality takes longer because the parent chain must process and finalize batches. The challenge protocol that backs this guarantee is described in [BoLD](/how-arbitrum-works/bold/gentle-introduction.mdx).
- **Cost considerations**: Posting batches to the parent chain incurs fees, potentially increasing transaction costs.

## How it works

> ### Part 1: Soft finality—the sequencer feed in detail

Reviewer: ```suggestion
### Part 1: Soft finality—the Sequencer feed in detail
```

#### 1. How a transaction becomes a feed message

1. You send a transaction. It doesn't go straight to the feed; it lands in a queue inside the Sequencer.

> 2. The Sequencer's main loop pulls your transaciton off the queue.

Reviewer: ```suggestion
2. The Sequencer's main loop pulls your transaction off the queue.
```

> 3. It hands the transaciton to the execution engine, which runs it and builds a new L2 block.

Reviewer: ```suggestion
3. It hands the transaction to the execution engine, which runs it and builds a new L2 block.
```

2. It stamps a `SequencerNumber` on it—a permanent "this is slot `N`" claim that no other transaction can ever take.
3. It includes the `BlockHash` so anyone can re-run the transaction themselves and check they get the same result.
4. It adds a `Signature` that only the Sequencer's key could produce.
5. A listener receives the envelope and now knows: the slot, the expected outcome, and proof of who wrote it.

#### 1c. The signature is what makes "soft" trust-minimal

1. The Sequencer gathers the fields that matter (chain ID, sequencer number, block hash, delayed-message count, raw transaction bytes) and glues them together.
2. It prepends a fixed label, "Arbitrum Nitro Feed:", so a signature for one chain can't be reused on another.
3. It hashes that whole blob and signs the hash with its private key.
4. The client receives the message, rebuilds the exact same blob and hash on its own.
5. The client checks the signature against that hash. If it matches, the message genuinely came from the Sequencer.
6. Because the Sequencer signed it, any attempt to later contradict slot `N` is permanent, public proof of cheating.

#### 1d. What the feed cannot guarantee yet

> 1. So far your transaction only lives in the Sequencer's memory and in listerners' caches—nothing is on Ethereum yet.

Reviewer: ```suggestion
1. So far your transaction only lives in the Sequencer's memory and in listeners' caches—nothing is on Ethereum yet.
```

> 2. If the Sequencer vanished right now, the ordering could still be lost or shuffled, because nothing durable recorded it.

Reviewer: ```suggestion
2. If the Sequencer vanished right now, the ordering could still be lost or shuffled, because no durable record of it exists.
```

3. The Sequencer could also just ignore your transaction and never sequence it at all.
4. To guard against being ignored, you can instead push your transaction through Ethereum's "delayed inbox," which forces eventual inclusion.
5. These gaps (durability + censorship) are exactly what posting a batch to L1 in Part 2 fixes.

### Part 2: Hard finality: Stage A—the Sequencer inbox commitment

#### 2a. The `BatchPoster` lifecycle

> 1. A timer fires, and the `BatchPoster` wwakes up to consider posting a new batch.

Reviewer: ```suggestion
1. A timer fires, and the `BatchPoster` wakes up to consider posting a new batch.
```

#### 3a. The `InboxReader` watches L1 at a chosen level

1. The `InboxReader` is the node's "eyes on Ethereum," constantly watching for posted batches.
2. The operator tells it how cautious to be: `latest` (fastest, can be reorged), `safe` (~30s lag), or `finalized` (~13 min, rock solid).
3. For `safe/finalized`, it asks L1 for the latest block at the confidence level and reads only up to there.
4. It translates that L1 block into "how many L2 messages are covered" via `GetSafeMsgCount` / `GetFinalizedMsgCount`.
5. When you query the L2 with "finalized," the node answers using only data Ethereum has already finalized.

#### 3b. `InboxTracker` writes the batch and emits a feed `Confirm`

1. The `InboxReader` spots a new batch event on L1 and hands it to the `InboxTracker`.
2. The `InboxTracker` records the batch in its local database.
3. To stay safe against a small L1 reorg, it doesn't confirm the just-posted batch—it confirms the one **before** the previous batch.
4. It sends a special "Confirm" message over the feed carrying that confirmed sequence number.
5. Feed-only clients see this and learn: these messages are now backed by Ethereum, not just the Sequencer's signature.

> ### Part 4: Hard finality, Stage B—Rollup assertion confirmation

Reviewer: ```suggestion
### Part 4: Hard finality: Stage B—Rollup assertion confirmation
```

#### 4a. Two protocols coexist in this repo

1. L1 inclusion proved the **data** exists, but not that the resulting L2 **state** is correct—that's a separate guarantee.
2. Validators ("bonders") are the ones who vouch for the state correctness on L1.

> 3. A wrapper checks the onchain Rollup contract at runtime and picks the dispute protocol the chain is using (the interactive legacy, or BoLD).

Reviewer: ```suggestion
3. A wrapper checks the onchain Rollup contract at runtime and picks the dispute protocol the chain is using (the interactive legacy protocol, or BoLD).
```

> 3. A wrapper checks the onchain rollupo contract at runtime and picks the dispute protocol the chain is using (the interactive legacy, or BoLD).

Reviewer: ```suggestion
3. A wrapper checks the onchain Rollup contract at runtime and picks the dispute protocol the chain is using (the interactive legacy, or BoLD).
```

4. Both protocols expose the same two signals—"latest bonded" and "latest confirmed"—so the rest of the node doesn't care which one is running.

#### 4b. The poll loop (legacy)

> 

Reviewer: ```suggestion
1. On a repeating timer, the legacy bonder checks the Rollup contract.
```
